utils/misc.mli

```diff
@@ -331,3 +331,137 @@ module type HookSig = sig
 end
 
 module MakeHooks : functor (M : sig type t end) -> HookSig with type t = M.t
+
+
+module Magic_number : sig
+  (** a typical magic number is "Caml1999O011"; it is formed of an
+      alpha-numeric prefix, here Caml1990O, followed by a version,
+      here 011. The prefix identifies the kind of the versioned data:
+      here the O indicates that it is the magic number for .cmo files.
+
+      All magic numbers have the same byte length, [magic_length], and
+      this is important for users as it gives them the number of bytes
+      to read to obtain the byte sequence that should be a magic
+      number. Typical user code will look like:
+      {|
+        let ic = open_in_bin path in
+        let magic =
+          try really_input_string ic Magic_number.magic_length
+          with End_of_line -> ... in
+        match Magic_number.parse magic with
+        | Error parse_error -> ...
+        | Ok info -> ...
+      |}
+
+      A given compiler version expects one specific version for each
+      kind of object file, and will fail if given an unsupported
+      version. Because versions grow monotonically, you can compare
+      the parsed version with the expected "current version" for
+      a kind, to tell whether the wrong-magic object file comes from
+      the past or from the future.
+
+      An example of code block that expects the "currently supported version"
+      of a given kind of magic numbers, here [Cmxa], is as follows:
+      {|
+        let ic = open_in_bin path in
+        begin
+          try Magic_number.(expect_current Cmxa (get_info ic)) with
+          | Parse_error error -> ...
+          | Unexpected error -> ...
+        end;
+        ...
+      |}
+
+      Parse errors distinguish inputs that are [Not_a_magic_number str],
+      which are likely to come from the file being completely
+      different, and [Corrupted str], raised by headers that are the
+      (possibly empty) prefix of a valid magic number.
+
+      Unexpected errors correspond to valid magic numbers that are not
+      the one expected, either because it corresponds to a different
+      kind, or to a newer or older version. The helper function
+      [explain] will generate a textual explanation of the mismatch
+      between the expectation and the actual header, for use in error
+      messages.
+
+      @since 4.05.0
+  *)
+
+  (* This API could be improved in the following ways:
+
+     - It is conceivable that future compiler versions might support
+       several versions of a given object file format at the same time
+       (for example to provide some transition period). Then the
+       current_* API would be insufficient, as a range or set of
+       versions would be supported. *)
```

Comment:
That's very optimistic and hypothetical. ;)

Reply:
I got rid of this comment.